Update openpilot warp and add failing self-test

haraschax opened the pull request:
Had to add all of the additional code that happens in openpilot, for it fail in the same way.

Currently it fails the self-test as follows (only happens when jitting, tinygrad without jit works fine):
```
batman@workstation-harald:~/xx/openpilot/tinygrad_repo$ python test/external/external_openpilot_image_warp.py
enqueue 1357.24 ms -- total run 1357.25 ms
enqueue 817.33 ms -- total run 817.34 ms
enqueue   5.60 ms -- total run   5.81 ms
enqueue   0.47 ms -- total run   0.68 ms
Traceback (most recent call last):
  File "/home/batman/xx/openpilot/tinygrad_repo/test/external/external_openpilot_image_warp.py", line 177, in <module>
    assert mismatch_percent < mismatch_percent_tol, f"input mismatch percent {mismatch_percent} exceeds tolerance {mismatch_percent_tol}"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: input mismatch percent 50.000762939453125 exceeds tolerance 0.01
```

All of the issues seem to come from editing the big buffer and passing it back in. If I add .clone() on line 73, the self-tests pass, but it's not clear to me why this is needed. That buffer should not need to be copied? If anything the other output buffer should be copied, but doing that doesn't fix it.

haraschax commented:
I added some realizes  and contiguous calls where it felt like it made sense, but not sure if correct.

I added the clone so the self-test passes.

Finally I added a pickle save and load test, which fails with:
```
enqueue   0.51 ms -- total run   0.63 ms
Traceback (most recent call last):
  File "/home/batman/xx/openpilot/tinygrad_repo/test/external/external_openpilot_image_warp.py", line 198, in <module>
    out2 = update_imgs(*(test_inputs_small + test_inputs_small2))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 338, in __call__
    ret = self.captured(input_buffers, var_vals)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 209, in __call__
    self._jit_cache = apply_graph_to_jit(self.jit_cache, input_buffers, var_vals, max_batch_size=JIT_BATCH_SIZE.value)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 64, in apply_graph_to_jit
    if len(current_batch) > 0: flush_batch()
                               ^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 31, in flush_batch
    graph_runner = current_batch_devs[0].graph(current_batch, input_rawbuffers, var_vals)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 14, in __init__
    self.devices = list(set(cast(HCQCompiled, d) for ji in jit_cache for d in [Device[cast(Buffer, x).device] for x in ji.bufs]))
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 14, in <genexpr>
    self.devices = list(set(cast(HCQCompiled, d) for ji in jit_cache for d in [Device[cast(Buffer, x).device] for x in ji.bufs]))
                                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 14, in <listcomp>
    self.devices = list(set(cast(HCQCompiled, d) for ji in jit_cache for d in [Device[cast(Buffer, x).device] for x in ji.bufs]))
                                                                                      ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'device'
Exception ignored in: <function HCQGraph.__del__ at 0x7f192cf05e40>
Traceback (most recent call last):
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 225, in __del__
    for dev in self.devices: self.last_timeline[dev][0].wait(self.last_timeline[dev][1])
               ^^^^^^^^^^^^
AttributeError: 'HCQGraph' object has no attribute 'devices'
```

geohot commented:
Still failing linter, mixed 2 and 4 spaces, tests should be minimal and test only what you want.

Additionally, if you put the test not in external, it will run in our CI and we will prevent regressions.

haraschax commented:
I fixed the lint.

I added the full implementation of how it's used in openpilot, I have no confidence a subset of this can guarantee the full function works. I encountered many different issues while making this.

It also cannot be added to CI yet, it currently still fails to run the pickle after it is load/saved:
```
enqueue   0.51 ms -- total run   0.63 ms
Traceback (most recent call last):
  File "/home/batman/xx/openpilot/tinygrad_repo/test/external/external_openpilot_image_warp.py", line 198, in <module>
    out2 = update_imgs(*(test_inputs_small + test_inputs_small2))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 338, in __call__
    ret = self.captured(input_buffers, var_vals)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 209, in __call__
    self._jit_cache = apply_graph_to_jit(self.jit_cache, input_buffers, var_vals, max_batch_size=JIT_BATCH_SIZE.value)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 64, in apply_graph_to_jit
    if len(current_batch) > 0: flush_batch()
                               ^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/engine/jit.py", line 31, in flush_batch
    graph_runner = current_batch_devs[0].graph(current_batch, input_rawbuffers, var_vals)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 14, in __init__
    self.devices = list(set(cast(HCQCompiled, d) for ji in jit_cache for d in [Device[cast(Buffer, x).device] for x in ji.bufs]))
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 14, in <genexpr>
    self.devices = list(set(cast(HCQCompiled, d) for ji in jit_cache for d in [Device[cast(Buffer, x).device] for x in ji.bufs]))
                                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 14, in <listcomp>
    self.devices = list(set(cast(HCQCompiled, d) for ji in jit_cache for d in [Device[cast(Buffer, x).device] for x in ji.bufs]))
                                                                                      ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'device'
Exception ignored in: <function HCQGraph.__del__ at 0x7f192cf05e40>
Traceback (most recent call last):
  File "/home/batman/xx/openpilot/tinygrad/runtime/graph/hcq.py", line 225, in __del__
    for dev in self.devices: self.last_timeline[dev][0].wait(self.last_timeline[dev][1])
               ^^^^^^^^^^^^
AttributeError: 'HCQGraph' object has no attribute 'devices'
```

In the meantime I while try to make minimal repros for the issues I encountrered.
